File: server/application/models/user/userao.php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

class UserAo extends CI_Model {
	
	var $TYPE_ADMIN = 0;
	var $TYPE_USER = 1;
	public function __construct(){
		parent::__construct();
		$this->load->model('user/userDb','userDb');
	}
	
	public function search($dataWhere,$dataLimit){
		return $this->userDb->search($dataWhere,$dataLimit);
	}
	
	public function get($userId){
		return $this->userDb->get($userId);
	}
	
	public function del($userId){
		return $this->userDb->del($userId);
	}
	
	public function add($data){
		//检查是否有重名
		$result = $this->userDb->getByName($data['name']);
		if( $result['code'] != 0 )
			return $result;
		$user = $result['data'];
		if( count($user) != 0 )
			return array(
				'code'=>1,
				'msg'=>'存在重复的用户名',
				'data'=>''
			);
		
		//添加用户
		$data['password'] = sha1($data['password']);
		return $this->userDb->add($data);
	}
	
	public function modType($userId,$type){
		$data = array();
		$data['type'] = $type;
		return $this->userDb->mod($userId,$data);
	}
	
	public function modPassword($userId,$password){
		$data = array();
		$data['password'] = sha1($password);
		return $this->userDb->mod($userId,$data);
	}
	
	public function modPasswordByOld($userId,$oldPassword,$newPassword){
		//检查是否有重名
		$result = $this->userDb->getByIdAndPass($userId,sha1($oldPassword));
		if( $result['code'] != 0 )
			return $result;
		$users = $result['data'];
		if( count($users) == 0 )
			return array(
				'code'=>1,
				'msg'=>'原密码错误',
				'data'=>''
			);
		
		//修改密码
		$data = array();
		$data['password'] = sha1($newPassword);
		return $this->userDb->mod($userId,$data);
	}
}

File: server/models/user/userao.go
package user

import (
	"crypto/sha1"
	"fmt"
	. "github.com/fishedee/language"
	. "github.com/fishedee/web"
	"io"
	. "mymanager/models/common"
)

type userAoModel struct {
	Model
	UserDb UserDbModel
}

func (this *userAoModel) CheckMustVaildPassword(password string, passwordHash string) {
	if this.getPasswordHash(password) != passwordHash {
		Throw(1, "密码不正确")
	}
}

func (this *userAoModel) getPasswordHash(password string) string {
	sha1er := sha1.New()
	io.WriteString(sha1er, password)
	dataHash := fmt.Sprintf("%x", sha1er.Sum(nil))
	return dataHash
}

func (this *userAoModel) Search(user User, pageInfo CommonPage) Users {
	return this.UserDb.Search(user, pageInfo)
}

func (this *userAoModel) Get(userId int) User {
	return this.UserDb.Get(userId)
}

func (this *userAoModel) GetByName(name string) []User {
	return this.UserDb.GetByName(name)
}

func (this *userAoModel) Del(userId int) {
	this.UserDb.Del(userId)
}

func (this *userAoModel) Add(user User) {
	//检查是否有重名
	userInfo := this.UserDb.GetByName(user.Name)
	if len(userInfo) != 0 {
		Throw(1, "存在重复的用户名")
	}

	//添加用户
	user.Password = this.getPasswordHash(user.Password)
	this.UserDb.Add(user)
}

func (this *userAoModel) ModType(userId int, userType int) {
	user := User{
		Type: userType,
	}
	this.UserDb.Mod(userId, user)
}

func (this *userAoModel) ModPassword(userId int, password string) {
	user := User{
		Password: this.getPasswordHash(password),
	}
	this.UserDb.Mod(userId, user)
}

func (this *userAoModel) ModPasswordByOld(userId int, oldPassword string, newPassword string) {
	//检查原密码是否正确
	userInfo := this.UserDb.Get(userId)
	this.CheckMustVaildPassword(oldPassword, userInfo.Password)

	//修改密码
	this.ModPassword(userId, newPassword)
}
